unified-runtime/source/adapters/cuda/command_buffer.hpp

```diff
@@ -122,142 +93,39 @@ struct kernel_command_handle : ur_exp_command_buffer_command_handle_t_ {
-                                                WaitNodes),
-        DecomposedNodes(std::move(DecomposedNodes)) {}
-  CommandType getCommandType() const noexcept override {
-    return CommandType::MemBufferFill;
-  }
+// Command handle that can be returned from command append entry-points.
+// The type of the command is specified by a CommandType field, with
+// additional command-type-specific data stored in the CommandData enum.
+struct ur_exp_command_buffer_command_handle_t_ {
+  using command_data_type_t =
+      std::variant<null_command_data, kernel_command_data, fill_command_data>;
 
-  // If this fill command was decomposed into multiple nodes, this vector
-  // contains all of those nodes in the order they were added to the graph.
-  // Currently unused but will be required for updating in future.
-  std::vector<CUgraphNode> DecomposedNodes;
-};
+  ur_exp_command_buffer_command_handle_t_(
+      CommandType Type, ur_exp_command_buffer_handle_t CommandBuffer,
+      CUgraphNode Node, CUgraphNode SignalNode,
+      const std::vector<CUgraphNode> &WaitNodes,
+      command_data_type_t Data = null_command_data{})
+      : CommandBuffer(CommandBuffer), Node(Node), SignalNode(SignalNode),
+        WaitNodes(WaitNodes), Type(Type), CommandData(Data) {}
 
-struct usm_prefetch_command_handle : ur_exp_command_buffer_command_handle_t_ {
-  usm_prefetch_command_handle(ur_exp_command_buffer_handle_t CommandBuffer,
-                              CUgraphNode Node, CUgraphNode SignalNode,
-                              const std::vector<CUgraphNode> &WaitNodes)
-      : ur_exp_command_buffer_command_handle_t_(CommandBuffer, Node, SignalNode,
-                                                WaitNodes) {}
-  CommandType getCommandType() const noexcept override {
-    return CommandType::USMPrefetch;
-  }
-};
+  // Parent UR command-buffer.
+  ur_exp_command_buffer_handle_t CommandBuffer;
+  // Node created in graph for the command.
+  CUgraphNode Node;
+  // An optional EventRecordNode that's a successor of Node to signal
+  // dependent commands outwith the command-buffer.
+  CUgraphNode SignalNode;
+  // Optional list of EventWait Nodes to wait on commands from outside of the
+  // command-buffer.
+  std::vector<CUgraphNode> WaitNodes;
 
-struct usm_advise_command_handle : ur_exp_command_buffer_command_handle_t_ {
-  usm_advise_command_handle(ur_exp_command_buffer_handle_t CommandBuffer,
-                            CUgraphNode Node, CUgraphNode SignalNode,
-                            const std::vector<CUgraphNode> &WaitNodes)
-      : ur_exp_command_buffer_command_handle_t_(CommandBuffer, Node, SignalNode,
-                                                WaitNodes) {}
-  CommandType getCommandType() const noexcept override {
-    return CommandType::USMAdvise;
-  }
+  CommandType Type;
+  std::variant<null_command_data, kernel_command_data, fill_command_data>
+      CommandData;
```

Comment:
```suggestion
  command_data_type_t CommandData;
```
Since we already defined the alias.